### PR TITLE
Added `transform_to_data_extent()`, improved `transform()`

### PR DESCRIPTION
**Pull request by LucaMarconato**
# New API:
- Added utils function `transform_to_data_extent()`, useful to create a converted with downscaled images, where all the -transformations are applied and where 1 pixel = 1 physical unit.

# Improved robustness and tests of transformations
- Added `maintain_positioning` argument to`SpatialData.transform_to_coordinate_system()` and `SpatialData.transform_element_to_coordinate_system()`.
- Fixed bug: incorrect offset translation vector for raster data when calling the two above functions, `transform()` was not affected)
- Fixed bug: `transform()` on multiscale raster was returning the offset translation for the most low res scale, not scale0

Most importantly:
- `transform()` now requires the argument `to_coordinate_system` instead of `transformation` (unless `maintain_positioning` is `True`); in fact, non-explicit target coordinate systems may lead to unintuitive results when an element is aligned to multiple coordinate systems. This change is communicated to the user by an informative exception; in some cases backward compatibility could be ensured so only an informative warning is given.

Note:
The above bugs were not affecting the rest of the codebase because `query()`, `aggregate()`, ... make either use of `transform()` for non-raster data, either operate on the transformations directly, without transforming raster data.

### CODE REVIEW

File: src/spatialdata/_core/spatialdata.py [hunk 1]
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
...
         t = get_transformation_between_coordinate_systems(self, element, target_coordinate_system)
-        transformed = transform(element, t, maintain_positioning=False)
-        remove_transformation(transformed, remove_all=True)
-        set_transformation(transformed, Identity(), target_coordinate_system)
-
+        if maintain_positioning:
+            transformed = transform(element, transformation=t, maintain_positioning=maintain_positioning)
+        else:
+            d = get_transformation(element, get_all=True)
+            assert isinstance(d, dict)
```
Comment (melonora) [inline src/spatialdata/_core/spatialdata.py:516]: Idem here, assert is not required I believe.
```suggestion
```

File: src/spatialdata/_core/spatialdata.py [hunk 2]
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
...
+                d[target_coordinate_system] = t
+                to_remove = True
+            transformed = transform(
+                element, to_coordinate_system=target_coordinate_system, maintain_positioning=maintain_positioning
+            )
+            if to_remove:
+                del d[target_coordinate_system]
+        if not maintain_positioning:
+            d = get_transformation(transformed, get_all=True)
+            assert isinstance(d, dict)
```
Comment (melonora) [inline src/spatialdata/_core/spatialdata.py:528]:
```suggestion
```

File: src/spatialdata/_core/spatialdata.py [hunk 3]
```diff
@@ -490,29 +490,72 @@ def transform_element_to_coordinate_system(
...
+            t = list(d.values())[0]
+            remove_transformation(transformed, remove_all=True)
+            set_transformation(transformed, t, target_coordinate_system)
+        else:
+            # When maintaining positioning is true, and if the element has a transformation to target_coordinate_system
+            # (this may not be the case because it could be that the element is not directly mapped to that coordinate
+            # system), then the transformation to the target coordinate system is not needed # because the data is now
+            # already transformed; here we remove such transformation.
+            d = get_transformation(transformed, get_all=True)
+            assert isinstance(d, dict)
```
Comment (melonora) [inline src/spatialdata/_core/spatialdata.py:539]:
```suggestion
```

File: src/spatialdata/_core/operations/rasterize.py [hunk 4]
```diff
@@ -448,7 +448,8 @@ def _(
     if schema in (Labels2DModel, Labels3DModel):
         kwargs = {"prefilter": False, "order": 0}
     elif schema in (Image2DModel, Image3DModel):
-        kwargs = {}
+        kwargs = {"order": 0}
+        # kwargs = {}
```
Comment (melonora) [inline src/spatialdata/_core/operations/rasterize.py:452]:
```suggestion
```

File: src/spatialdata/_core/operations/transform.py [hunk 5]
```diff
@@ -132,124 +135,192 @@ def _prepend_transformation(
...
         assert raster_translation is None
-        if maintain_positioning:
-            to_prepend = transformation.inverse()
+        to_prepend = transformation.inverse() if maintain_positioning else Identity()
     else:
         raise TypeError(f"Unsupported type {type(element)}")
+    assert isinstance(to_prepend, BaseTransformation)
 
     d = get_transformation(element, get_all=True)
     assert isinstance(d, dict)
```
Comment (melonora) [inline src/spatialdata/_core/operations/transform.py:168]:
```suggestion
```

File: src/spatialdata/_core/operations/transform.py [hunk 6]
```diff
@@ -132,124 +135,192 @@ def _prepend_transformation(
...
+    data: SpatialElement,
+    transformation: BaseTransformation | None,
+    maintain_positioning: bool,
+    to_coordinate_system: str | None,
+) -> tuple[BaseTransformation, str | None]:
+    from spatialdata.transformations import BaseTransformation, get_transformation
+
+    if not maintain_positioning:
+        d = get_transformation(data, get_all=True)
+        assert isinstance(d, dict)
```
Comment (melonora) [inline src/spatialdata/_core/operations/transform.py:192]:
```suggestion
```

**Review by melonora — COMMENTED** [review]
Very quick first pass:)

File: tests/core/operations/test_transform.py [hunk 7]
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
...
+def _postpone_transformation(
+    sdata: SpatialData, from_coordinate_system: str, to_coordinate_system: str, transformation: BaseTransformation
+):
+    for element in sdata._gen_elements_values():
+        d = get_transformation(element, get_all=True)
+        assert isinstance(d, dict)
+        assert len(d) == 1
+        t = d[from_coordinate_system]
+        sequence = Sequence([t, transformation])
+        set_transformation(element, sequence, to_coordinate_system)
```
Comment (melonora) [inline tests/core/operations/test_transform.py:130]: Perhaps better to move this to a `_test_utils.py` in src/spatialdata.

File: tests/core/operations/test_transform.py [hunk 8]
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
...
-    affine = _get_affine(small_translation=False)
-    padded = transform(
-        transform(sdata, affine, maintain_positioning=False), affine.inverse(), maintain_positioning=False
-    )
-    _unpad_rasters(padded)
-    # raise NotImplementedError("TODO: plot the images")
-    # raise NotImplementedError("TODO: compare the transformed images with the original ones")
+def _remove_coordinate_system(sdata: SpatialData, coordinate_system: str):
+    for element in sdata._gen_elements_values():
+        remove_transformation(element, coordinate_system)
```
Comment (melonora) [inline tests/core/operations/test_transform.py:135]: Same as above

File: tests/core/operations/test_transform.py [hunk 9]
```diff
@@ -117,65 +118,88 @@ def _unpad_rasters(sdata: SpatialData) -> SpatialData:
...
-    # TODO: unpad the image
-    # raise NotImplementedError("TODO: compare the transformed images with the original ones")
+def _compare_extents(sdata0: SpatialData, sdata1: SpatialData):
+    e0 = get_extent(sdata0)
+    e1 = get_extent(sdata1, coordinate_system="transformed_back")
 
+    # we are not pixel perfect here, but "0.1-pixel perfect", I think this is due to an equivalent bug than then one
+    # that appears with rasterize (https://github.com/scverse/spatialdata/issues/165), probably originating from
+    # ndinterp(), which is used both by rasterize() and by transform()
+    assert all(np.allclose(e0[k], e1[k], atol=0.1) for k in set(e0.keys()).union(e1.keys()))
```
Comment (melonora) [inline tests/core/operations/test_transform.py:145]: same as above

**Review by melonora — APPROVED** [review]
For record, the asserts were due to mypy complaining that a `BaseTransformation` does not have `keys`, while there is already an assert for the case that `get_all` is `True`. The other option other than the assert would be to ignore these mypy errors.

I think the PR can be approved, but I do think for future and with the refactor more comments can be added here and there to explain the code to make it more clear. Right now without the explanation on zoom it could be hard to follow or maybe that is just me:)

File: src/spatialdata/_core/data_extent.py [hunk 10]
```diff
@@ -289,9 +286,7 @@ def _(e: GeoDataFrame, coordinate_system: str = "global", exact: bool = True) ->
             coordinate_system=coordinate_system,
             extent=extent,
         )
-    t = get_transformation(e, to_coordinate_system=coordinate_system)
-    assert isinstance(t, BaseTransformation)
-    transformed = transform(e, t)
+    transformed = transform(e, to_coordinate_system=coordinate_system)
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/data_extent.py:289]: Done.

File: src/spatialdata/_core/data_extent.py [hunk 11]
```diff
@@ -368,8 +359,10 @@ def _compute_extent_in_coordinate_system(
         max_coordinate=max_coordinates,
     )
     df = pd.DataFrame(corners.data, columns=corners.axis.data.tolist())
-    points = PointsModel.parse(df, coordinates={k: k for k in axes})
-    transformed_corners = pd.DataFrame(transform(points, transformation).compute())
+    d = get_transformation(element, get_all=True)
+    assert isinstance(d, dict)
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/data_extent.py:363]: you are right, I have removed it

File: src/spatialdata/_core/operations/_utils.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    from spatialdata.transformations.transformations import BaseTransformation, Scale, Sequence, Translation
+
+    sdata = sdata.filter_by_coordinate_system(coordinate_system=coordinate_system)
+    # calling transform_to_coordinate_system will likely decrease the resolution, let's use rasterize() instead
+    sdata_vector = SpatialData(shapes=dict(sdata.shapes), points=dict(sdata.points))
+    sdata_raster = SpatialData(images=dict(sdata.images), labels=dict(sdata.labels))
+    sdata_vector_transformed = sdata_vector.transform_to_coordinate_system(coordinate_system)
+
+    de = get_extent(sdata, coordinate_system=coordinate_system)
+    de_axes = tuple(de.keys())
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/_utils.py:67]: Done, also for `de_axes`

File: src/spatialdata/_core/operations/transform.py (hunk 6 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/transform.py:192]: makes mypy complain

File: src/spatialdata/_core/operations/transform.py (hunk 5 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/transform.py:168]: makes mypy complain

File: src/spatialdata/_core/spatialdata.py (hunk 1 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:516]: makes mypy complain

File: src/spatialdata/_core/spatialdata.py (hunk 2 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:528]: makes mypy complain

File: src/spatialdata/_core/spatialdata.py (hunk 3 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/spatialdata.py:539]: makes mypy complain

File: tests/core/operations/test_transform.py (hunk 9 above)
Comment (LucaMarconato) [reply tests/core/operations/test_transform.py:145]: Good point, actually I have renamed it to `are_extents_equal() -> bool` and moved it to `data_extent.py`

File: tests/core/operations/test_transform.py (hunk 8 above)
Comment (LucaMarconato) [reply tests/core/operations/test_transform.py:135]: moved to utils function `remove_transformations_to_coordinate_system()`

File: tests/core/operations/test_transform.py (hunk 7 above)
Comment (LucaMarconato) [reply tests/core/operations/test_transform.py:132]: I would keep this function here, at least for the moment, because I think it's better for the user to do this operation manually. In this case we assume that len(d) == 1, but in more complex scenario you may have an intricate graph of transformations, and this function could compromise some alignments to other coordinate systems.

**Comment by LucaMarconato** [comment]
Thanks for the review. I have addressed all the comments (and left 3 conversations open to see if you agree with my proposal). I'll fix the docs, otherwise if you re-approve we are good to merge.

File: src/spatialdata/_core/operations/transform.py [hunk 13]
```diff
@@ -104,24 +107,24 @@ def _transform_raster(
...
     raster_translation: Translation | None,
     maintain_positioning: bool,
+    to_coordinate_system: str | None = None,
 ) -> None:
-    """Prepend a transformation to an element.
-
-    After an element has been transformed, this method is called to eventually prepend a particular transformation to
-    the existing transformations of the element. The transformation to prepend depends on the type of the element (
-    raster vs non-raster) and on the maintain_positioning flag.
+    """Adjust the transformations of an element after it has been transformed.
```
Comment (kevinyamauchi) [inline src/spatialdata/_core/operations/transform.py:118]: Please add a more clear docstring. I am not sure what "adjustments" are being made

File: src/spatialdata/_core/operations/_utils.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+def transform_to_data_extent(
+    sdata: SpatialData,
+    coordinate_system: str,
+    target_unit_to_pixels: float | None = None,
+    target_width: float | None = None,
+    target_height: float | None = None,
+    target_depth: float | None = None,
+) -> SpatialData:
+    """
+    Transform the spatial data to match the data extent, keeping the positioning, and making all transformations equal.
```
Comment (kevinyamauchi) [inline src/spatialdata/_core/operations/_utils.py:21]: Can this be another operation than scale? If not, I think we should say scale. I think "tranform" is ambiguous. 

Also what is meant by "making all transformations equal"?

File: tests/core/operations/test_transform.py [hunk 15]
```diff
@@ -459,14 +481,111 @@ def test_map_coordinate_systems_long_path(full_sdata):
...
+    for element in full_sdata._gen_elements_values():
+        # from spatialdata.transformations.operations import _build_transformations_graph
+        #
+        # g = _build_transformations_graph(full_sdata)
+        # import networkx as nx
+        #
+        # nx.draw(g, with_labels=True)
+        # import matplotlib.pyplot as plt
+        #
+        # plt.show()
```
Comment (kevinyamauchi) [inline tests/core/operations/test_transform.py:550]: Do we need these commented lines still?

File: tests/core/operations/test_transform.py [hunk 16]
```diff
@@ -459,14 +481,111 @@ def test_map_coordinate_systems_long_path(full_sdata):
...
+            images=dict(full_sdata.images),
+            labels=dict(full_sdata.labels),
+            points=dict(full_sdata.points),
+            shapes=dict(full_sdata.shapes),
+            table=full_sdata.table,
+        )
+        temp["transformed_element"] = transformed_element
+        t = get_transformation_between_coordinate_systems(temp, temp["transformed_element"], "multi_hop_space")
+        a = t.to_affine_matrix(input_axes=("x",), output_axes=("x",))
+        d = get_transformation(transformed_element, get_all=True)
```
Comment (kevinyamauchi) [inline tests/core/operations/test_transform.py:564]: Can you please use more informative variable names?

**Review by kevinyamauchi — APPROVED** [review]
I think this looks good. I left some clarifying questions below. Please feel free to merge once you've addressed them.

**Comment by melonora** [comment]
I also believe it can be merged after @kevinyamauchi comments are adressed.

File: src/spatialdata/_core/operations/transform.py (hunk 13 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/transform.py:117]: I renamed the function to  `_set_transformation_for_transformed_elements()` and fixed the docstring. In short, transformed elements don't have a transformation and this needs to be put inside. The transformation is not just the old transformation(s), but needs some "adjustments", which I now explain better.

File: src/spatialdata/_core/operations/_utils.py (hunk 14 above)
Comment (LucaMarconato) [reply src/spatialdata/_core/operations/_utils.py:21]: In `Transform the spatial data to match the data extent, keeping the positioning, and making all transformations equal.`

The `Transform` is referring to the application of the `transform()` operation. The transformation being applied can be anything, not just scale (affine, etc).

The returned object will be such that if we have the point coordinate (x, y), this will correspond to the pixel (y, x). The problem is that in this way we lose the positioning to the original data, so we still need to add scale and translations to make the returned data align with the original data. 

I will add these information to the docstring.

Comment (LucaMarconato) [reply src/spatialdata/_core/operations/_utils.py:21]: I have also added an extra argument, called `maintain_positioning: bool = True`, which if false sets all the transformations to `Identity()`. I explained this argument in the docstring.

File: tests/core/operations/test_transform.py (hunk 15 above)
Comment (LucaMarconato) [reply tests/core/operations/test_transform.py:550]: removed.

File: tests/core/operations/test_transform.py (hunk 16 above)
Comment (LucaMarconato) [reply tests/core/operations/test_transform.py:564]: I agree, here it looks ugly. In general often in the code I use `t` for transformation, `a` for affine and `d` for the dict of transformations returned by `get_transformation(..., get_all=True)` (which is followed by `assert isinstance(d, dict)` to address `mypy`. So I think the `d` letter is ok.

**Comment by LucaMarconato** [comment]
Thank you both for the review, I have addressed the review comments. 

I also added an argument `maintain_positioning: bool = True` to `transform_to_data_extent()`, which can be more intuitive for the user (I explained this in the docstring).

I also added some extra tests for checking better the results of `transform_to_data_extent()`, and test the new argument.

Merging now 🥳
